genrl/deep/common/trainer.py

```diff
@@ -1,516 +1,472 @@
-                    }
-                )
-
-            if self.render:
-                self.env.render()
-
-            if self.save_interval != 0 and epoch % self.save_interval == 0:
-                self.save(epoch * self.agent.batch_size)
-
-        self.env.close()
-        self.logger.close()
+from abc import ABC
+from typing import Any, List, Optional, Type, Union
+
+import gym
+import numpy as np
+import torch
+
+from ...environments import VecEnv
+from .buffers import PrioritizedBuffer, ReplayBuffer
+from .logger import Logger
+from .utils import save_params, set_seeds
+
+
+class Trainer(ABC):
+    """
+    Base Trainer class. To be inherited specific usecases.
+
+    :param agent: Algorithm object
+    :param env: Environment
+    :param logger: Logger object
+    :param x_axis: Key plotted on x_axis 
+    :param buffer: Buffer Object
+    :param off_policy: Is the algorithm off-policy?
+    :param save_interval: Model to save in each of these many timesteps
+    :param render: Should the Environment render
+    :param max_ep_len: Max Episode Length
+    :param distributed: (True if distributed training is enabled, else
+False (To be implemented))
+    :param ckpt_log_name: Model checkpoint name
+    :param steps_per_epochs: Steps to take per epoch?
+    :param epochs: Total Epochs to train for
+    :param device: Device to train model on
+    :param log_interval: Log important params every these many steps
+    :param batch_size: Size of batch
+    :param seed: Set seed for reproducibility
+    :param deterministic_actions: Take deterministic actions during training.
+    :type agent: object
+    :type env: object
+    :type logger: object
+    :type x_axis: str
+    :type buffer: object
+    :type off_policy: bool
+    :type save_interval: int
+    :type render: bool
+    :type max_ep_len: int
+    :type distributed: bool
+    :type ckpt_log_name: string
+    :type steps_per_epochs: int
+    :type epochs: int
+    :type device: string
+    :type log_interval: int
+    :type batch_size: int
+    :type seed: int
+    :type deterministic_actions: bool
+    """
+
+    def __init__(
+        self,
+        agent: Any,
+        env: Union[gym.Env, VecEnv],
+        log_mode: List[str] = ["stdout"],
+        x_axis: str = "timestep",
+        buffer: Union[Type[ReplayBuffer], Type[PrioritizedBuffer]] = None,
+        off_policy: bool = False,
+        save_interval: int = 0,
+        render: bool = False,
+        max_ep_len: int = 1000,
+        distributed: bool = False,
+        ckpt_log_name: str = "experiment",
+        steps_per_epoch: int = 4000,
+        epochs: int = 10,
+        device: Union[torch.device, str] = "cpu",
+        log_interval: int = 10,
+        evaluate_episodes: int = 500,
+        logdir: str = "logs",
+        batch_size: int = 50,
+        seed: Optional[int] = None,
+        deterministic_actions: bool = False,
+        transform: bool = None,
+        history_length: int = 4,
```

Comment:
Once you merge with the upstream, these should go. Keep the changes from the upstream wherever it's conflicting unless it's directly related to the logging functionality.